### DISCUSSION TITLE
Implement functionality equivalent to `eslint-plugin-react-refresh`

### DISCUSSION THREAD

**Discussion by GunseiKPaseri** [biomejs/biome#2548] [Ideas]
from https://github.com/biomejs/biome/discussions/3#discussioncomment-9177559
# react-refresh
[react-refresh](https://www.npmjs.com/package/react-refresh) is an official package of react that provides bundler with a Fast Refresh feature that allows real-time updates of running React apps.
In addition to [react-native](https://www.npmjs.com/package/react-native), [next.js](https://www.npmjs.com/package/next), [@vitejs/plugin-react](https://www.npmjs.com/package/@vitejs/plugin-react) and [@parcel/runtime-react-refresh](https://www.npmjs.com/package/@parcel/runtime-react-refresh) are often used behind the scenes in development environments.
It is a so-called hot reload, but it is characterized by the fact that only edited components can be updated minute by minute while maintaining the state.
For more detailed information, please refer to the documentation of each bundler.
- [React Native](https://reactnative.dev/docs/fast-refresh)
- [next.js](https://nextjs.org/docs/architecture/fast-refresh)
- [Remix](https://remix.run/docs/en/main/discussion/hot-module-replacement)
- [@vitejs/plugin-react](https://www.npmjs.com/package/@vitejs/plugin-react#consistent-components-exports)
- [Gatsby](https://www.gatsbyjs.com/docs/reference/local-development/fast-refresh/#how-it-works)

# eslint-plugin-react-refresh
[eslint-plugin-react-refresh](https://www.npmjs.com/package/eslint-plugin-react-refresh) is a plugin also linked from [@vitejs/plugin-react](https://www.npmjs.com/package/@vitejs/plugin-react) that defines rules to make react-refresh function properly.

One rule, `react-refresh/only-export-components`, contains the following
- No simultaneous export of components and non components (in JSX)
  - Components are judged based on whether they are PascalCases or not. 
    - also need [react/jsx-pascal-case](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-pascal-case.md)
      - [`useNamingConvention`](https://biomejs.dev/linter/rules/use-naming-convention/) (See https://github.com/biomejs/biome/discussions/3#discussioncomment-9177700)
  - [allowExportNames option](https://github.com/ArnaudBarre/eslint-plugin-react-refresh?tab=readme-ov-file#allowexportnames-v044)
  - [allowConstantExport option](https://github.com/ArnaudBarre/eslint-plugin-react-refresh?tab=readme-ov-file#allowconstantexport-v040)
- All-uppercase function export is considered an error when not using direct named export (ex `const CMS = () => <></>; export { CMS }`
  - Error because the confirmation function does not support it. react-refresh works.
- No `export *` (in JSX)
  - Error because the confirmation function does not support it.
  - **[noReExportAll](https://biomejs.dev/linter/rules/no-re-export-all) in biome**
- No anonymous export (in JSX)
  - Anonymous functions are not supported and the value of such states will disappear on update.
    - Reference: [Hot Module Replacement | Remix](https://remix.run/docs/en/main/discussion/hot-module-replacement#named-function-components)
  - like [import/no-anonymous-default-export](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-anonymous-default-export.md)
  - Worst case scenario, [noDefaultExport](https://biomejs.dev/linter/rules/no-default-export/) can be substituted.
- No unexported components (in JSX)

In addition, the following, although not implemented, would be nice to have
- no Class Components
  - like [eslint-plugin-react-prefer-function-component](https://www.npmjs.com/package/eslint-plugin-react-prefer-function-component)

**I would like your input on implementing these in Biome.**
- Do similar rules exist inside or outside of Biome?
- Can it be more abstract?
- What should naming be like? I'm not good at English😖.

**Comment by heyitsarpit** [comment]
this is the one plugin blocking me from updating everything to biome right now.

**Reply by kikoane** [reply]
how do you handle css-in-js ?

**Comment by gvillo** [comment]
is there an issue ticket for this? it would be awesome to add this

**Reply by ematipico** [reply]
We don't add tickets unless there's people interested in doing the work, or the maintainers and core contributors can do the work

**Comment by GunseiKPaseri** [comment]
@ematipico Can I try it?

**Reply by ematipico** [reply]
Of course! Please create an issue (task) and assign it to you. Feel free to leave any questions in the issue

**Reply by GunseiKPaseri** [reply]
Thanks! https://github.com/biomejs/biome/issues/3560

**Comment by GunseiKPaseri** [comment]
https://github.com/biomejs/biome/issues/3560 has been merged and published!
https://github.com/biomejs/biome/releases/tag/cli%2Fv1.9.2
